common/core/core_session.c

```diff
@@ -0,0 +1,482 @@
+ * initialization.
+ *
+ * @param server_message The buffer containing the server's message, which
+ * includes the server's random public key, session age, and device ID.
+ * @return true if the server's key is successfully received and processed.
+ * @return false otherwise.
+ */
+static bool session_receive_server_key(uint8_t *server_message);
+
+/**
+ * @brief Initializes the curve parameters for the session.
+ */
+static void session_curve_init();
+
+/**
+ * @brief Derives the server's public key from the extended public key.
+ *
+ * @return true if the server's public key is successfully derived.
+ * @return false otherwise.
+ */
+static bool derive_server_public_key();
+
+/**
+ * @brief Verifies the session signature using the provided payload and
+ * signature.
+ *
+ * @param payload The payload data to verify.
+ * @param payload_size The size of the payload.
+ * @param signature The signature to verify.
+ * @return true if the signature is valid.
+ * @return false otherwise.
+ */
+static bool verify_session_signature(const uint8_t *payload,
+                                     const size_t payload_size,
+                                     const uint8_t *signature);
+
+/**
+ * @brief Derives the session initialization vector (IV) using the server's and
+ * device's random public points.
+ */
+static void derive_session_iv();
+
+/**
+ * @brief Derives the session key using ECDH multiplication.
+ *
+ * @return true if the session key is successfully derived.
+ * @return false otherwise.
+ */
+static bool derive_session_key();
+
+/**
+ * @brief Derives both the session IV and session key.
+ *
+ * @return true if both the session IV and session key are successfully derived.
+ * @return false otherwise.
+ */
+static bool derive_session_iv_and_session_key();
+
+/**
+ * @brief Appends a signature to the payload.
+ *
+ * @param payload The payload to append the signature to.
+ * @param payload_size The size of the payload.
+ */
+static void session_append_signature(uint8_t *payload,
+                                     const size_t payload_size);
+
+/**
+ * @brief Generates random keys for the session.
+ *
+ * @param random The buffer to store the generated random key.
+ * @param random_public The buffer to store the generated random public key.
+ * @param random_public_point The curve point to store the random public key
+ * point.
+ * @return true if the random keys are successfully generated.
+ * @return false otherwise.
+ */
+static bool session_get_random_keys(uint8_t *random,
+                                    uint8_t *random_public,
+                                    curve_point random_public_point);
+
+/**
+ * @brief Sends the device key as part of the session initialization.
+ *
+ * @param payload The buffer to store the payload data.
+ * @return true if the device key is successfully sent.
+ * @return false otherwise.
+ */
+static bool session_send_device_key(uint8_t *payload);
```

Comment:
```suggestion
static bool session_generate_device_key(uint8_t *payload);
```
the function is not sending the keys but only storing it in the ip buffer,
also update the description accordingly